feat(middleware): added manual file type option

josh18 opened the pull request:
Adds a manual file type so that users can instruct Karma which method to use when loading a file without an extension (such as Google font files)

Example:
```
files: [
    {
        pattern: 'https://fonts.googleapis.com/icon?family=Material+Icons',
        type: 'css'
    }
]
```

Fix for https://github.com/karma-runner/karma/issues/2824

First time working karma so let me know if there are any changes I need to make.

josh18 commented:
I signed it!

josh18 commented:
1. Sure thing
2. As this is the default is this necessary? Or are you just thinking of adding it to the validation in (3)?
3. Will do, what would I use to log it? Just something like `log.warn()`?
Another thing I was just thinking of is if it should be `fileExt === '.css' || fileType === 'css'` (can't override file extension) or `(fileExt === '.css' && !fileType) || fileType === 'css'` (fileType always overrides extension)

appsforartists commented on lib/middleware/karma.js:
```diff
@@ -193,12 +194,12 @@ var createKarmaMiddleware = function (
+            if (fileExt === '.html' || fileType === 'html') {
               scriptTags.push(util.format(LINK_TAG_HTML, filePath))
               continue
             }
```
On line 208, `scriptType` should be `fileType || SCRIPT_TYPE[fileExt] || 'text/javascript'`.  I think that would let us strip the code out of #2834 and just merge its tests.

appsforartists replied:
GitHub sadly wouldn't let me comment on the correct line.

josh18 commented:
Made some changes, let me know what you think.

- File type overrides extension
- Added script file types (including esModule)
- Added file type validation
- Added file type documentation
  - Renamed one instance of `Possible values` -> `Possible Values` for consistency
  - Removed extra white space (done automatically by my editor, not sure if you want to keep this change)

appsforartists commented on lib/middleware/karma.js:
```diff
@@ -32,9 +33,17 @@ var CROSSORIGIN_ATTRIBUTE = 'crossorigin="anonymous"'
+  'html',
+  'js',
+  'dart',
+  'esModule'
```
This is total bikeshedding, but this could be more simply called `module`.  It removes the only camel-cased option, and makes the API more consistent with the browser: `type: 'module'`.

josh18 replied:
That's fine I was just copying from the previous PR but I think `'module'` is better. Fixed now

appsforartists commented on docs/config/01-configuration-file.md:
```diff
@@ -425,7 +425,7 @@ httpsServerOptions: {
 **CLI:** `--log-level debug`
 
-**Possible values:**
+**Possible Values:**
```
Pull requests usually follow the mantra "one idea -> one PR".  This correction should probably have been a separate PR.

josh18 replied:
Makes sense, sorry about that. Reverted now. Should I squash all my commits?

dignifiedquire commented:
Thank you :octocat:
